R/teal_data_module-eval_code.R

```diff
@@ -0,0 +1,124 @@
+setOldClass("teal_data_module")
+
+#' Evaluate the code in the qenv environment
+#' @name eval_code
+#' @description
+#' Given code is evaluated in the `qenv` environment of `teal_data` reactive defined in `teal_data_module`.
+#' @param object (`teal_data_module`)
+#' @inheritParams teal.code::eval_code
+#' @return Returns a `teal_data_module` object.
+#' @importMethodsFrom teal.code eval_code
+#' @importFrom methods setMethod
+NULL
+
+#' @rdname eval_code
+#' @export
+#'
+#' @examples
+#' tdm <- teal_data_module(
+#'   ui = function(id) div(id = shiny::NS(id)("div_id")),
+#'   server = function(id) {
+#'     shiny::moduleServer(id, function(input, output, session) {
+#'       shiny::reactive(teal_data(IRIS = iris))
+#'     })
+#'   }
+#' )
+#' \dontrun{
+#' eval_code(tdm, "IRIS <- subset(IRIS, Species == 'virginica')")
+#' }
+setMethod("eval_code", signature = c("teal_data_module", "character"), function(object, code) {
+  teal_data_module(
+    ui = function(id) {
+      ns <- NS(id)
+      object$ui(ns("mutate_inner"))
+    },
+    server = function(id) {
+      moduleServer(id, function(input, output, session) {
+        teal_data_rv <- object$server("mutate_inner")
+
+        if (!is.reactive(teal_data_rv)) {
+          stop("The `teal_data_module` must return a reactive expression.", call. = FALSE)
+        }
+
+        eventReactive(teal_data_rv(),
+          {
+            data <- tryCatch(teal_data_rv(), error = function(e) e)
+
+            if (inherits(data, "teal_data")) {
+              eval_code(data, code)
+            } else if (inherits(data, "error")) {
+              data
+            } else {
+              validate(
+                need(
+                  FALSE,
+                  paste(
+                    sep = "\n",
+                    "Error when executing `teal_data_module`:",
+                    paste0(
+                      "It must always return a reactive with `teal_data`, it returns object of class(es): ",
```

Comment:
As in `srv_teal_with_splash`, I believe this error message is for the app dev, not for the app user. It should shop up in the console, not in the app.

Reply:
It depends. Let's say we have teal_data_module that reads csv files only and an app user select e.g. pdf file then we probably need to expose error message for app user.